Fixed Issue #127

shivenmian opened the pull request:
I've come up with a system to handle errors as in Issue #127. It will show dialogs and give the respective error that has occurred.

For HTTP errors, it gives a dialog with the Error code too. For others, it gives a dialog with the type of error and the cause.

When network is unavailable, it gives a dialog to turn on network in Settings.

shivenmian commented:
@mananwason @creativepsyco Please check this and tell me if something more is required.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/activities/MainActivity.java:
```diff
@@ -72,7 +80,33 @@ protected void onCreate(Bundle savedInstanceState) {
+        else {
+            downloadProgress.setVisibility(View.GONE);
+            AlertDialog.Builder builder = new AlertDialog.Builder(this);
+            builder.setTitle("Network not available")
```
Strings should go inside `Strings.xml`

creativepsyco commented on app/src/main/java/org/fossasia/openevent/activities/MainActivity.java:
```diff
@@ -72,7 +80,33 @@ protected void onCreate(Bundle savedInstanceState) {
+            download.downloadVersions();
+        }
+        else {
+            downloadProgress.setVisibility(View.GONE);
```
what happens if a dialog is already present.

mananwason commented on app/src/main/java/org/fossasia/openevent/api/APIClient.java:
```diff
@@ -37,11 +44,61 @@ public APIClient() {
+                            }
+                        } else if (cause.getKind().equals(RetrofitError.Kind.HTTP)) {
+                            try {
+                                Log.d("Type", "HTTPError");
```
Add tags to all the logs

creativepsyco commented on app/src/main/java/org/fossasia/openevent/activities/MainActivity.java:
```diff
@@ -72,7 +80,33 @@ protected void onCreate(Bundle savedInstanceState) {
+                    .setNeutralButton("Cancel", new DialogInterface.OnClickListener() {
+                        @Override
+                        public void onClick(DialogInterface dialog, int which) {
+
```
You should probably move this to a different class and pass the current activity as an argument instead of accessing it via `MainActivity.this.finish()`

shivenmian replied:
Why?

The purpose of MainActivity.this.finish() is to close the app as soon as I go to Settings, otherwise download won't happen even after switching WiFi/Data on.

creativepsyco replied:
You should implement a NetworkConnectivityChange receiver to monitor connectivity. And don't do it in the AndroidManifest, do it in the application onCreate. See http://stackoverflow.com/questions/12274997/why-broadcastreceiver-works-even-when-app-is-in-background and https://gist.github.com/aniXification/5896231

mananwason commented on app/src/main/java/org/fossasia/openevent/events/ErrorDialogEvent.java:
```diff
@@ -0,0 +1,14 @@
+
+    public String s1;
+    public String s2;
+    public ErrorDialogEvent(String s1, String s2){
```
Write descriptive names for both the strings

creativepsyco commented on app/src/main/java/org/fossasia/openevent/activities/MainActivity.java:
```diff
@@ -72,7 +80,33 @@ protected void onCreate(Bundle savedInstanceState) {
+                            MainActivity.this.finish();
+                        }
+                    })
+                    .setNeutralButton("Cancel", new DialogInterface.OnClickListener() {
```
There is a dialog without a neutral button as well, again captions etc. go inside `strings.xml`

creativepsyco commented on app/src/main/java/org/fossasia/openevent/activities/MainActivity.java:
```diff
@@ -72,7 +80,33 @@ protected void onCreate(Bundle savedInstanceState) {
+        if(networkInfo!=null && networkInfo.isConnected()){
+            download.downloadVersions();
+        }
+        else {
```
Indentation, brace position.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/activities/MainActivity.java:
```diff
@@ -58,6 +65,7 @@
     private int currentMenuItemId;
     private final static String STATE_FRAGMENT = "stateFragment";
 
+
```
empty line?

creativepsyco commented on app/src/main/java/org/fossasia/openevent/activities/MainActivity.java:
```diff
@@ -72,7 +80,33 @@ protected void onCreate(Bundle savedInstanceState) {
+                        @Override
+                        public void onClick(DialogInterface dialog, int which) {
+
+                        }
```
This method does nothing? Then why is this here.

shivenmian replied:
It does. This will be called when network is not present. The app won't start if network is not there.

creativepsyco replied:
I am talking about the `onClick` listener for the neutral button.

shivenmian replied:
Oh. Yeah that does nothing, you just go back to the activity. I'll remove it.

creativepsyco replied:
@shivenmian please look at my comment above, I am talking about the `onClick` empty function, it is a dialog button click listener and has nothing to do with network conditions.

shivenmian replied:
Yeah, I'll put null for it instead of onClick. I'd kept the Cancel button should the user not go to Settings and stay offline.

creativepsyco replied:
yes that would be better.

mananwason commented on app/src/main/java/org/fossasia/openevent/activities/MainActivity.java:
```diff
@@ -327,7 +360,8 @@ public void noInternet(NoInternetEvent event) {
     public void cantDownload(CantDownloadEvent event) {
         downloadProgress.setVisibility(View.GONE);
-        Snackbar.make(mainFrame, getString(R.string.cantDownload), Snackbar.LENGTH_LONG).show();
+
```
Remove extra lines

creativepsyco commented on app/src/main/java/org/fossasia/openevent/activities/MainActivity.java:
```diff
@@ -359,4 +393,21 @@ public void onMicrolocationsDownloadDone(MicrolocationDownloadEvent event) {
+        AlertDialog.Builder builder = new AlertDialog.Builder(this);
+        builder.setTitle("Error")
+                .setMessage(event.s1 + ": " + event.s2)
+                .setNeutralButton("OK", new DialogInterface.OnClickListener() {
```
You should pass `null` as the listener if nothing is being done, again strings go into a separate file.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/api/APIClient.java:
```diff
@@ -37,11 +44,61 @@ public APIClient() {
+                                ErrorDialogEvent errEvent = new ErrorDialogEvent("Network Error", String.valueOf(cause.getCause()));
+                                OpenEventApp.postEventOnUIThread(errEvent);
+                                Log.d("ErrorType", String.valueOf(cause.getResponse().getStatus()));
+                            } catch (NullPointerException e) {
```
Why are you catching null pointer exception?

shivenmian replied:
I had this as an error once so added it. I'll remove it, it's not needed now.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/api/APIClient.java:
```diff
@@ -37,11 +44,61 @@ public APIClient() {
+                            Log.d("Type", "Other Error");
+                            Log.d("ErrorType", String.valueOf(cause.getResponse().getStatus()));
+                            ErrorDialogEvent errEvent = new ErrorDialogEvent("Other Error", String.valueOf(cause.getCause()));
+                            OpenEventApp.postEventOnUIThread(errEvent);
```
Too many repeated statements here, you can use a switch case and fix this up. use chaining (or builder pattern) to fix up generating an event based on the arguments involved like `event.setType().setCause().setCode()` etc.

shivenmian replied:
I don't think a switch case is beneficial because I've used .equals() instead of ==, and the former is more preferred than the latter. 

I've chained stuff though, dropping a PR soon :)

creativepsyco replied:
I believe that switch-case is pretty efficient in terms of enums, which is the case here, nevertheless I feel switch case is more readable and easier to maintain than an if-then-else.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/api/APIClient.java:
```diff
@@ -37,11 +44,61 @@ public APIClient() {
+                })
                 .build();
         openEventAPI = adapter.create(OpenEventAPI.class);
+
```
empty line

creativepsyco commented on app/src/main/java/org/fossasia/openevent/api/APIClient.java:
```diff
@@ -37,11 +44,61 @@ public APIClient() {
 
     public OpenEventAPI getOpenEventAPI() {
         return openEventAPI;
+
```
empty line.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/api/processor/EventListResponseProcessor.java:
```diff
@@ -46,7 +46,6 @@ public void run() {
 
     }
 
-
```
unexpected deletion?

shivenmian replied:
Fixed.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/api/processor/SpeakerListResponseProcessor.java:
```diff
@@ -57,6 +57,6 @@ public void run() {
     public void failure(RetrofitError error) {
         // Do something with failure, raise an event etc.
-        OpenEventApp.getEventBus().post(new SponsorDownloadEvent(false));
+        OpenEventApp.getEventBus().post(new SpeakerDownloadEvent(false));
```
This apparently is a different fix, not related to showing up dialogs, why is this here, part of this commit?

shivenmian replied:
Yeah this isn't related to showing up of dialogs, this is possibly a small typo @mananwason did, so I fixed it too.

creativepsyco replied:
In either case, should not be part of your commit.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/events/ErrorDialogEvent.java:
```diff
@@ -0,0 +1,14 @@
+ */
+public class ErrorDialogEvent {
+
+    public String s1;
```
Better naming please.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/activities/MainActivity.java:
```diff
@@ -327,7 +360,8 @@ public void noInternet(NoInternetEvent event) {
-        Snackbar.make(mainFrame, getString(R.string.cantDownload), Snackbar.LENGTH_LONG).show();
+
+
     }
```
extra lines?

shivenmian replied:
I'd added something in this function and deleted it later. I'll remove them.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/activities/MainActivity.java:
```diff
@@ -187,7 +221,6 @@ private void syncComplete() {
     private void downloadFailed() {
         downloadProgress.setVisibility(View.GONE);
         Snackbar.make(mainFrame, getString(R.string.download_failed), Snackbar.LENGTH_LONG).show();
-
```
unexpected deletion?

shivenmian replied:
Oops, did it accidentaly.

shivenmian replied:
It doesn't matter anyway, there wasn't anything there, just extra lines.

shivenmian replied:
Fixed.

creativepsyco commented:
Requires heavy cleanup.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/activities/MainActivity.java:
```diff
@@ -327,7 +360,8 @@ public void noInternet(NoInternetEvent event) {
     @Subscribe
     public void cantDownload(CantDownloadEvent event) {
```
Why are there 2 events: `CantDownloadEvent` & `ErrorDialogEvent` I don't there should be 2 events for 1 point of failure, unless you have 2 different UI components or some background processing, you should consider using only one and remove the other.

shivenmian replied:
cantDownload was already there, didn't touch this. I'll remove it, it isn't needed.

creativepsyco commented:
Also commit message is missing, can use a better description than `Updated`

shivenmian commented:
@creativepsyco @mananwason I've fixed stuff, if there's anything more let me know.

For the ErrorDialogs, I created another function in the ErrorDialog.java and used it to shorten code.

mananwason commented on app/src/main/java/org/fossasia/openevent/api/processor/VersionApiProcessor.java:
```diff
@@ -104,7 +107,7 @@ public void run() {
 
     @Override
     public void failure(RetrofitError error) {
-        OpenEventApp.getEventBus().post(new CantDownloadEvent());
```
Shouldn't you fire a ErrorDialogEvent here?

shivenmian replied:
It gives two dialogs. I tried it.

creativepsyco replied:
I think you should not depend on APIClient for your implementation. The code should be as decoupled as possible. You should continue firing the error or at least log it, but whether the MainActivity chooses to subscribe to it or whether it has shown a network error dialog, should be upto it.

Also I think that instead of setting an Error Handler on the Retrofit adapter you should piggy back on the event that the processor generates and transform the event into something that you can show (the dialog in your case). The reason I am saying this is because the error generated from the Processor is local, has more contextual information, you know exactly where the error is coming from, while for the Retrofit adapter it will cause the handler to fire for all global error issues. I don't think that should be the approach.

mananwason commented on app/src/main/java/org/fossasia/openevent/api/processor/SpeakerListResponseProcessor.java:
```diff
@@ -47,7 +47,7 @@ public void run() {
 
-                OpenEventApp.postEventOnUIThread(new SpeakerDownloadEvent(true));
+                OpenEventApp.postEventOnUIThread(new SponsorDownloadEvent(true));
             }
```
I believe this should be SpeakerDownloadEvent Only. Please fix it.

shivenmian replied:
You asked me to change it to Sponsor :$

mananwason replied:
@creativepsyco asked you to change it in the failure callback.

shivenmian replied:
Yeah I realised.

shivenmian commented:
@mananwason @creativepsyco Fixed stuff. If there's anything more, please let me know.

To handle errors and tidy up code, I created a new function in ErrorDialog.java and used it throughout.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/activities/MainActivity.java:
```diff
@@ -86,6 +99,23 @@ protected void onCreate(Bundle savedInstanceState) {
 
     }
 
+    public void WiFiDialog(final Activity activity){
```
starting method name with capitals? Also should use a verb for function name, like `showXXXX` instead of a noun.

shivenmian replied:
Done.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/activities/MainActivity.java:
```diff
@@ -86,6 +99,23 @@ protected void onCreate(Bundle savedInstanceState) {
+                    public void onClick(DialogInterface dialog, int which) {
+                        Intent setNetworkIntent = new Intent(Settings.ACTION_SETTINGS);
+                        startActivity(setNetworkIntent);
+                        activity.finish();
```
I don't think you need to call activity.finish(), I know you are doing this to force download the information the next time the person comes back to the app, but I think that should be handled from a connectivity change receiver because network can change anytime and you need that.

shivenmian replied:
I tried keeping my network on and then running the app and switching it off while the download is in progress. It gives a Network Error (Socket time-out exception) everytime.

Instead of adding a BroadcastReceiver, why not just add this case ^ to my existent method?

creativepsyco replied:
We want to refresh the information whenever we have an active network connection and broadcast receivers allow us to do that.

shivenmian replied:
Oh okay, didn't realise you wanted this. Sure, will do.

creativepsyco replied:
I mean whether you implement it as part of a new GCI task or you don't implement it is up to you. But at least the finish statement makes it look it the problem is fixed. But its not :)

shivenmian replied:
I'll fix it. 

PS: I'm not doing GCI. I'm in college. GSoC :stuck_out_tongue:

creativepsyco commented on app/src/main/java/org/fossasia/openevent/activities/MainActivity.java:
```diff
@@ -359,4 +382,15 @@ public void onMicrolocationsDownloadDone(MicrolocationDownloadEvent event) {
     }
 
+    @Subscribe
+    public void ErrorDialog(ErrorDialogEvent event){
```
method names are improper, should be verbs and follow camel case notation. Please see the names of the functions within the file to get an idea.

shivenmian replied:
Fixed.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/events/ErrorDialogEvent.java:
```diff
@@ -0,0 +1,59 @@
+ */
+public class ErrorDialogEvent {
+
+    public String ErrorType;
```
class attributes starting with caps?

shivenmian replied:
Fixed.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/events/ErrorDialogEvent.java:
```diff
@@ -0,0 +1,59 @@
+    public static final String TYPE = "RetrofitError Type";
+    public static final String ERROR_CODE = "Error Code";
+
+    public ErrorDialogEvent(String s1, String s2){
```
descriptive string please, not `s1` `s2`

shivenmian replied:
Fixed.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/api/APIClient.java:
```diff
@@ -37,6 +48,15 @@ public APIClient() {
+                .setErrorHandler(new ErrorHandler() {
+                    @Override
+                    public Throwable handleError(RetrofitError cause) {
+                        ErrorDialogEvent errEvent = new ErrorDialogEvent(null, null);
```
Hmm. Why are you creating an object and then passing it to your handler?

shivenmian replied:
Fixed.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/api/APIClient.java:
```diff
@@ -37,6 +48,15 @@ public APIClient() {
+                    @Override
+                    public Throwable handleError(RetrofitError cause) {
+                        ErrorDialogEvent errEvent = new ErrorDialogEvent(null, null);
+                        errEvent.ErrorHandlerFunction(cause, errEvent);
```
Here you are calling a member function and passing the member as an argument. This is definitely not right.

shivenmian replied:
That's because if you look at ErrorHandlerFunction, I'm changing errorType and errorDesc based on the type of error (i.e cause). That's why I first call ErrorDialogEvent and pass null null, then change the member's parameters using the second line.

creativepsyco replied:
You can just create an `ErrorDialogEvent` from the function itself right?

shivenmian replied:
I haven't looked at all the other issues you pointed, will do in a bit. But for this one, I've done this, look at these three images: [This](http://i.imgur.com/Bw41D12.png), [this](http://i.imgur.com/8AgNo3h.png) and [this](http://i.imgur.com/lebbW8J.png).

The first one is of the handleError function in APIClient, the next two are of HandleErrorDialogEvent.java (I changed the name).

Is this okay?

creativepsyco replied:
I think we should drop the errorHandler in retrofit client, use the one from the process see [above](https://github.com/fossasia/open-event-android/pull/128#discussion_r48522034)

shivenmian replied:
Hmm. I'll give it a try.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/api/APIClient.java:
```diff
@@ -37,6 +48,15 @@ public APIClient() {
+                    public Throwable handleError(RetrofitError cause) {
+                        ErrorDialogEvent errEvent = new ErrorDialogEvent(null, null);
+                        errEvent.ErrorHandlerFunction(cause, errEvent);
+                        OpenEventApp.postEventOnUIThread(new ErrorDialogEvent(errEvent.ErrorType, errEvent.ErrorDesc));
```
You already created an object in the lines above. I am not sure what you are doing here.

shivenmian replied:
Fixed.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/api/APIClient.java:
```diff
@@ -37,6 +48,15 @@ public APIClient() {
+                        ErrorDialogEvent errEvent = new ErrorDialogEvent(null, null);
+                        errEvent.ErrorHandlerFunction(cause, errEvent);
+                        OpenEventApp.postEventOnUIThread(new ErrorDialogEvent(errEvent.ErrorType, errEvent.ErrorDesc));
+                        return new Exception(cause);
```
`RetrofitError` extends from `RuntimeException` I don't believe you need to create a new object here.

shivenmian replied:
The reason I created the object was because all the error dialogs differed only in the type of error and the description, and that can be changed using one function and if-else / switch. Once you get that, you can simply post it.

creativepsyco replied:
I don't think we should do that because the error will be posted for all API calls, and this API Client is shared everywhere. The if-else or switch will create only 1 object so I don't really understand what's your point here:

> The reason I created the object errEvent was because all the error dialogs differed only in the type of error and the description

shivenmian replied:
Yeah, the error dialogs only differ in the type of error and the description. Check out HandleErrorDialogEvent. The constructor takes in two strings.

creativepsyco replied:
if the constructor is available you can just create the object right? Why do you need to initialise it with `null` and then pass it. The thing I don't get is that whats the point of having a member function when you are using it as a static helper method. Plus your datastructure returns nothing and you create another copy of the object from it, which is completely baffling. Objects are *NOT* cheap.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/api/APIClient.java:
```diff
@@ -28,7 +34,12 @@
 
+    public static final String TYPE = "RetrofitError Type";
+
+    public static final String ERROR_CODE = "Error Code";
```
unused strings

shivenmian replied:
Fixed.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/api/APIClient.java:
```diff
@@ -28,7 +34,12 @@
 
     private final OpenEventAPI openEventAPI;
 
+    public static final String TYPE = "RetrofitError Type";
```
unused strings.

shivenmian replied:
Fixed.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/events/ErrorDialogEvent.java:
```diff
@@ -0,0 +1,59 @@
+
+    public void ErrorHandlerFunction(RetrofitError cause, ErrorDialogEvent event){
+
+        if (cause.getKind().equals(RetrofitError.Kind.CONVERSION)) {
```
please use switch case, and make this return an `ErrorDialogEvent` object.

shivenmian replied:
Will do.

shivenmian replied:
I've made it return an ErrorDialogEvent object.

I tried Switch case with cause.getKind() being compared, but just as I thought, it didn't work. 

When I took cases to be compared as RetrofitError.Kind.CONVERSION etc, it gave an error saying "Constant Expression required". And it's apt. RetrofitError.Kind.CONVERSION is not an object so the error came.

Then I took the cases as merely CONVERSION, NETWORK, HTTP etc. The code worked without errors. But then I took the base url as a malformed JSON ([this](http://www.mocky.io/v2/5682128c120000993893a2c1)) and ran. Instead of conversion error, it gave HTTP 200 error.

The if-else had worked perfectly with the malformed JSON and it did give me conversion error. The problem with switch is that it uses '==' and this compares objects as a whole, while .equals() also compares parts. That's why switch won't work in this case as I expected and saw. Check [this](http://stackoverflow.com/questions/1643067/whats-the-difference-between-equals-and).

creativepsyco replied:
My IDE seems to be okay with this:
![image](https://cloud.githubusercontent.com/assets/547235/12030917/a98c8f5a-ae40-11e5-92b4-9f9b883d25b0.png)

>Then I took the cases as merely CONVERSION, NETWORK, HTTP etc. The code worked without errors. But then I took the base url as a malformed JSON (this) and ran. Instead of conversion error, it gave HTTP 200 error.

200 is success NOT error.

shivenmian replied:
Yeah, I know. I just mean to say that it gives the wrong dialog. It's supposed to give an error.

http://www.mocky.io/v2/5682128c120000993893a2c1

This is a malformed JSON. This should give a Conversion Error instead of an HTTP 200.

creativepsyco replied:
And I get 

![image](https://cloud.githubusercontent.com/assets/547235/12030961/5a377a04-ae41-11e5-890a-91777ea7e371.png) 
when I plug in http://www.mocky.io/v2/5682128c120000993893a2c1

I think it works, do you have more counterexamples?

shivenmian replied:
Err, was giving an error in my case. I'll try it once more. I had the exact same code, mind you.

creativepsyco replied:
I just pulled your branch in, don't think there are any differerences. Here's my diff, just some formatting issues:

https://gist.github.com/creativepsyco/ea33fd373d79f52d47a7

shivenmian replied:
I didn't really push since morning. I made a lot of changes. Do I push right now?

creativepsyco replied:
You can push anytime do take a look at the issues we discussed here, fix them and hopefully you learn something from the process.

shivenmian replied:
Yeah I've fixed most of what you've asked me to, myself. :)

I'm fixing that WiFi BroadcastReceiver thing and removing ErrorHandler from APIClient. Rest done. I'll push now, you can see.

creativepsyco commented:
@shivenmian Still a long way to go, lots of issues. Please fix. Thanks.

creativepsyco commented:
@shivenmian I believe switch case works fine. Please give it a shot! Also as far as switch case vs comparison branches (I don't know from where you believe that switch uses `==` and not `equals`) is concerned please take a look at the [JVM spec][1], I don't want to argue but if you are interested clearly from spec we can see that N no of comparisons (if-else-if) are better a single table lookup (switch-case). And internally byte code maps switch case labels as integers, and enums are perfect for this because they are sparse and uniformly distributed ordinals.


[1]: https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-3.html

shivenmian commented:
Switch-case optimises, no doubt, but in my case it gave an error, for which I tried to come up with a plausible explanation.

shivenmian commented:
@mananwason @creativepsyco Fixed stuff since morning. Just pushed. Apart from the TODOs I have written, if there are any other issues please let me know.

creativepsyco commented:
Apart from the above issues we discussed I don't think much is required. Just need to fix up your function names, variable names to align with Java Style Guide. Remove if-else in favour of switch, add a broadcast receiver to detect connectivity changes to trigger download (you can do this in a separate PR) instead of finishing the activity. Use Processor#failure() events instead of handling globally through Retrofit. Then I think this is good to go.

shivenmian commented:
You did check my latest [commit](https://github.com/shivenmian/open-event-android/commit/f5f6bc28342b0ede5fbaf02dc63ee9da7d0ceddf), right?

creativepsyco commented:
Yes I did. Still has the same issues. https://github.com/shivenmian/open-event-android/commit/f5f6bc28342b0ede5fbaf02dc63ee9da7d0ceddf

shivenmian commented:
Just the TODOs you mean. Okay. I'll fix them.

creativepsyco commented:
No not just the TODOs please see the issues we have gone over in this discussion thread and check your diff again after pushing. You can see all the changes here: https://github.com/fossasia/open-event-android/pull/128/files

shivenmian commented:
Fixed the lines stuff and the function names I believe. The comments on https://github.com/fossasia/open-event-android/pull/128/files show only the things I marked as TODO. I think that's the only thing left now @creativepsyco

shivenmian commented:
I have that NetworkBroadcastReceiver fix left. I have done the rest. @mananwason @creativepsyco 

Please tell me if any other fix left, apart from removing the extra URLs in the URL file. Those are tests. I'll do that in the end.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/api/APIClient.java:
```diff
@@ -3,11 +3,15 @@
 
+import retrofit.ErrorHandler;
 import retrofit.RestAdapter;
+import retrofit.RetrofitError;
```
remove the imports as well.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/api/APIClient.java:
```diff
@@ -29,6 +33,7 @@
     private final OpenEventAPI openEventAPI;
 
     public APIClient() {
+
```
remove the extra line.

creativepsyco commented:
ok only the TODOs left, I have added minor cleanup comments.

shivenmian commented:
Only one TODO is left. I converted to Switch-case, removed the error-handling from the RestAdapter. Only the NetworkBroadcastReceiver TODO is left.

@creativepsyco is the rest ok?

creativepsyco commented:
yes looks fine.

shivenmian commented:
@mananwason @creativepsyco Sorry for the long overdue. Was out the whole day. Just came back.

I've added the BroadcastReceiver, I think it works although I'm not fully sure. If there are any issues, please tell me.

creativepsyco commented on app/src/main/AndroidManifest.xml:
```diff
@@ -49,6 +49,7 @@
             android:enabled="true" />
 
         <receiver android:name=".Receivers.NotificationAlarmReceiver" />
+
```
Empty line

shivenmian commented:
Apart from the lines, is the rest fine?

I wanted to register the broadcastreceiver on the Manifest but then decided against it.

Is the rest working?

creativepsyco commented on app/src/main/java/org/fossasia/openevent/activities/MainActivity.java:
```diff
@@ -71,8 +80,9 @@ protected void onCreate(Bundle savedInstanceState) {
         downloadProgress.setIndeterminate(true);
-        DataDownload download = new DataDownload();
-        download.downloadVersions();
+        registerReceiver(new NetworkDialogHandler(), new IntentFilter(Context.CONNECTIVITY_SERVICE));
```
Receiver should be registered in Application#onCreate() because Activity can be created multiple times.

shivenmian replied:
registerReceiver is already in Mainactivity#onCreate(). Please check.

Or do you mean I should keep it in OpenEventApp#oncreate()? OpenEventApp extends Application.

creativepsyco replied:
Application not Activity

creativepsyco commented on app/src/main/java/org/fossasia/openevent/activities/MainActivity.java:
```diff
@@ -71,8 +80,9 @@ protected void onCreate(Bundle savedInstanceState) {
-        download.downloadVersions();
+        registerReceiver(new NetworkDialogHandler(), new IntentFilter(Context.CONNECTIVITY_SERVICE));
+        Intent intent = new Intent(Context.CONNECTIVITY_SERVICE);
+        sendBroadcast(intent);
```
why are you sending a broadcast here?

creativepsyco commented on app/src/main/java/org/fossasia/openevent/activities/MainActivity.java:
```diff
@@ -254,6 +281,36 @@ private void doMenuAction(int menuItemId){
+            isNetworkAvailable(context);
+        }
+
+        private boolean isNetworkAvailable(Context context){
```
this function is returning a `boolean` but you don't seem to using it anywhere

creativepsyco commented on app/src/main/java/org/fossasia/openevent/activities/MainActivity.java:
```diff
@@ -254,6 +281,36 @@ private void doMenuAction(int menuItemId){
+        private boolean isConnected = false;
+        @Override
+        public void onReceive(Context context, Intent intent) {
+            Log.d("NetNotif", "Received notif about net status");
```
Probably a better log message indicating what are the contents of the intent.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/activities/MainActivity.java:
```diff
@@ -254,6 +281,36 @@ private void doMenuAction(int menuItemId){
+                if(networkInfo.getState().equals(NetworkInfo.State.CONNECTED)){
+                    if(!isConnected){
+                        isConnected=true;
+                        DataDownload download = new DataDownload();
```
You should probably check if a data download is already in progress or not.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/activities/MainActivity.java:
```diff
@@ -254,6 +281,36 @@ private void doMenuAction(int menuItemId){
+                }
+            }
+            Log.d("NetNotif", "Not connected to Internet");
+            showNetworkDialog(MainActivity.this);
```
you cannot call view or activity related code from a BroadcastReceiver. This is wrong usage,

creativepsyco commented on app/src/main/java/org/fossasia/openevent/activities/MainActivity.java:
```diff
@@ -254,6 +281,36 @@ private void doMenuAction(int menuItemId){
+                    }
+                    return true;
+                }
+            }
```
where is the `else` branch.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/activities/MainActivity.java:
```diff
@@ -254,6 +281,36 @@ private void doMenuAction(int menuItemId){
+            showNetworkDialog(MainActivity.this);
+            isConnected=false;
+            return false;
+        }
```
Code is not formatted properly.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/api/Urls.java:
```diff
@@ -27,6 +27,6 @@
 
     public static final String BASE_GET_URL_ALT = "https://raw.githubusercontent.com/fossasia/open-event/master/testapi";
-
+    
```
unclean.

creativepsyco commented:
@shivenmian formatting issues + Broadcast receiver cannot communicate to any view or activity you should use event bus events for that. also please do a check so that downloading info is not called multiple times.

shivenmian commented:
@creativepsyco @mananwason Sorry for being a bit late. Was stuck up a bit lately.

I've implemented the broadcastreceiver and all the features you asked, it's working fine for me.

If there are any other issues, please tell me.

Oh, and Happy New Year! :smiley:

creativepsyco commented on app/src/main/AndroidManifest.xml:
```diff
@@ -49,6 +49,12 @@
             android:enabled="true" />
 
         <receiver android:name=".Receivers.NotificationAlarmReceiver" />
+        <receiver android:name=".Receivers.NetworkDialogHandler">
```
No need for the receiver to be declared here if you are registering it in the `onCreate` of the Application.

shivenmian replied:
No, I didn't register in the onCreate. I registered it statically.

creativepsyco replied:
When you register it in the Manifest it gets activated everytime the network changes, even if your app isn't working or killed by the system, by registering in the manifest you are infact making the phone do more work even when the app is not active. Do you know why wasn't it working? Its a broadcast receiver only getting triggered when the network changes. So your testing steps should be to switch off network, startup the app and then switch the network back on again.

shivenmian replied:
I'd done exactly that but I guess dynamic registration should be avoided or something, no? Most people do it in the manifest or so.

creativepsyco replied:
we need to do dynamic registration. A lot of people don't understand the difference, please make sure that you do.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/OpenEventApp.java:
```diff
@@ -1,11 +1,16 @@
 import com.squareup.otto.Bus;
 
+import org.fossasia.openevent.Receivers.NetworkDialogHandler;
+import org.fossasia.openevent.activities.MainActivity;
```
Why these imports are here?

shivenmian commented:
Fixed the lines and the redundant imports. Are there any other issues? Please tell me if so. @creativepsyco

creativepsyco commented on app/src/main/java/org/fossasia/openevent/activities/MainActivity.java:
```diff
@@ -71,10 +78,15 @@ protected void onCreate(Bundle savedInstanceState) {
-        download.downloadVersions();
-
         this.findViewById(android.R.id.content).setBackgroundColor(Color.LTGRAY);
+        ConnectivityManager connMgr = (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);
```
Why is this here? Shouldn't the processor be firing the `ShowNetworkDialogEvent`? Maybe the connection is weaker, maybe it takes forever to download the info. Or maybe there is a network timeout. You are just handling one situation here.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/events/ShowNetworkDialogEvent.java:
```diff
@@ -0,0 +1,23 @@
+ */
+public class ShowNetworkDialogEvent {
+
+    public Context context;
```
never cache context anywhere, it is a memory leak.

creativepsyco commented:
Hi @shivenmian I am unable to understand what's your approach. You should probably discuss why are you doing certain things over Gitter or IRC. For the me the ideal way to inform the user about network errors would be the activity just starts the Data download action, it doesn't care about whether the network is on or off. The processor that gets triggered later will throw an error if data download could not be successful, you can then check whether it is the network that is the problem or whether server is down and so on and fire the relevant events.

The network receiver will just fire an event telling whether network came back offline/online. The activity can listen to this event and restart the download if previously it didn't. You can track whether there is a download request in progress by using a boolean member variable in the activity.

Is that clear info for you? Please discuss this strategy, I feel commenting on the PR and you repeating your mistakes is a waste of my time and yours as well.

shivenmian commented:
My approach was to first check about Network presence and then call DataDownload, where the other errors are sorted out. That's because when Network isn't present, I intended on leading the user to the Settings page, while for other errors, the user just gets notified about them. Once network presence is confirmed, the other connection errors will be fired if they are present, but I intended to do a pre-check for Network presence because of the Settings thing.

shivenmian commented:
@creativepsyco Please check Gitter, I've explained what I've done in the code. There are a lot of comments as I didn't want to delete anything before being sure. I'll cleanup later.

Please tell me about any other issues.

creativepsyco commented on app/src/main/java/org/fossasia/openevent/activities/MainActivity.java:
```diff
@@ -359,4 +386,86 @@ public void onMicrolocationsDownloadDone(MicrolocationDownloadEvent event) {
+            }
+            case NETWORK: {
+                if(!(netinfo!=null && netinfo.isConnected())){ //Network not connected error
+                    OpenEventApp.postEventOnUIThread(new ShowNetworkDialogEvent(MainActivity.this));
```
Yes this is what was in my mind. But you can move the switch-case to a static function. like 
```java
Retrofit error = /* Obtained from somewhere */
errorType = NetworkUtils.getNetworkError(error)
if (errorType == null && !connected()) {
    // if network error dialog is not showing
    // show network connection dialog
} else {
    // display normal error dialog.
}
```

shivenmian commented:
@creativepsyco Please check this PR out and tell me if there are any other issues.

I'm a beginner at minimalising stuff so if there's something silly I've missed out, please tell me about it.

creativepsyco commented on app/src/main/AndroidManifest.xml:
```diff
@@ -49,6 +49,12 @@
+                <action android:name="android.net.conn.CONNECTIVITY_CHANGE" />
+            </intent-filter>
+        </receiver>
+
```
why is this still here?

shivenmian replied:
Oops. Removing it.

shivenmian commented:
Fixed @creativepsyco . Any other issues?

creativepsyco commented:
yes rest is fine.

creativepsyco commented:
Will merge thanks :+1:

creativepsyco commented:
Although there are a few more formatting issues (I will fix those) rest is fine, You got what I wanted to convey to you in the end which is great :) hopefully you continue to contribute to this project in the future. Cheers!

shivenmian commented:
Yeah I'm definitely contributing. Cheers!

creativepsyco commented:
Before committing you should reformat each file that you are changing to fix up formatting according to Android Studio Default style.

See option below, it will take care of lining up braces etc.
![image](https://cloud.githubusercontent.com/assets/547235/12116741/fb302356-b3f6-11e5-9a7e-b7dc67f0f123.png)

shivenmian commented:
Hmm. Thanks so much! :smile: :smile: :smile:
